lxd/metrics/api_rates.go

```diff
@@ -0,0 +1,93 @@
+package metrics
+
+import (
+	"net/url"
+	"sync/atomic"
+
+	"github.com/canonical/lxd/shared/entity"
+)
+
+// RequestStatus represents the a completed request status category
+type requestStatus string
+
+const (
+	serverError requestStatus = "server_error"
+	clientError requestStatus = "client_error"
+	success     requestStatus = "succeeded"
+)
+
+// PossibleRequestStatus is a slice that includes all the possible request status categories
+var PossibleRequestStatus = []requestStatus{
+	serverError,
+	clientError,
+	success,
+}
+
+func resolveStatusCode(statusCode int, asyncOperation bool) requestStatus {
+	if asyncOperation {
+		if statusCode == 400 {
+			return serverError
+		}
+		return success
+	}
+	if statusCode > 500 && statusCode < 600 {
+		return serverError
+	}
+	if statusCode > 400 && statusCode < 500 {
+		return clientError
+	}
+	return success
+}
+
+func resolveEntityType(url url.URL) entity.Type {
+	entityType, _, _, _, _ := entity.ParseURL(url)
```

Comment:
what entity type is returned if url is invalid?

Reply:
I'm not sure `entity.ParseURL` is fit for purpose here. It is used for specific LXD entities, e.g. an instance `/1.0/instances/c1?project=p1`. It will not work for requests that e.g. create an instance `POST /1.0/instances` since this URL does not reference a specific entity.

Reply:
@markylaing would it be useful/appropriate to update `entity.ParseURL` to recognise the top-level listing endpoints as return entityType, but an empty arg list?

Reply:
> what entity type is returned if url is invalid?

That would be a part of 'Fine tune entity_type assignment' so still have to approach this

Reply:
> @markylaing would it be useful/appropriate to update `entity.ParseURL` to recognise the top-level listing endpoints as return entityType, but an empty arg list?

We discussed this in stand up yesterday. I think I'd prefer to keep `entity.ParseURL` for parsing URLs of specific entities for the sake of simplicity with authorization. We discussed a couple of options:
1. Name all of our endpoints, and call `mux.CurrentRoute(r).GetName()`. @hamistao said that this would be too specific, but we could categorise these into endpoint groups.
2. Create another function for categorising URLs. It could call `entity.ParseURL` and fall back to a URL prefix.

I'm in favour of 1. since mux has already done the route parsing for us.

Reply:
Currently I have no preference between these options. But instead of 2, I think it would be simpler to just parse the endpoints in one dedicated function without `entity.ParseURL` (as the implementation currently stands) because, although `entity.ParseURL` would categorize most endpoints, it would still leave so many endpoints out that wouldn't end up helping much in terms of simplifying the code.

Reply:
@hamistao I had a brief discussion with Tom about this. The URL prefix should be sufficient for this, we mentioned simply using the second element of the URL path. So `/1.0/instances` will fall under `instances` (though there will need to be a special case here to capture `containers` and `vms`), `/1.0/storage` will fall under `storage` and so on.

Reply:
Then the current implementation should suffice, thanks :D
Did I miss this meeting by any chance? If so I am sorry, I did not see it on my calendar :/